benches/criterion.rs

```diff
@@ -0,0 +1,271 @@
+    };
+    for input in inputs.iter() {
+        let string = std::str::from_utf8(input).unwrap();
+        parser.parse(&options, string).unwrap();
+    }
+}
+
+fn frontends(c: &mut Criterion) {
+    let mut group = c.benchmark_group("front");
+    #[cfg(all(feature = "wgsl-in", feature = "serialize", feature = "deserialize"))]
+    group.bench_function("bin", |b| {
+        let inputs_wgsl = gather_inputs("tests/in", "wgsl");
+        let mut parser = naga::front::wgsl::Parser::new();
+        let inputs_bin = inputs_wgsl
+            .iter()
+            .map(|input| {
+                let string = std::str::from_utf8(input).unwrap();
+                let module = parser.parse(string).unwrap();
+                bincode::serialize(&module).unwrap()
+            })
+            .collect::<Vec<_>>();
+        b.iter(move || {
+            for input in inputs_bin.iter() {
+                bincode::deserialize::<naga::Module>(input).unwrap();
+            }
+        });
+    });
+    #[cfg(feature = "wgsl-in")]
+    group.bench_function("wgsl", |b| {
+        let inputs = gather_inputs("tests/in", "wgsl");
+        let mut parser = naga::front::wgsl::Parser::new();
+        b.iter(move || {
+            for input in inputs.iter() {
+                let string = std::str::from_utf8(input).unwrap();
+                parser.parse(string).unwrap();
+            }
+        });
+    });
+    #[cfg(feature = "spv-in")]
+    group.bench_function("spv", |b| {
+        let inputs = gather_inputs("tests/in/spv", "spv");
+        b.iter(move || {
+            let options = naga::front::spv::Options::default();
+            for input in inputs.iter() {
+                let spv =
+                    unsafe { slice::from_raw_parts(input.as_ptr() as *const u32, input.len() / 4) };
+                let parser = naga::front::spv::Parser::new(spv.iter().cloned(), &options);
+                parser.parse().unwrap();
+            }
+        });
+    });
+    #[cfg(feature = "glsl-in")]
+    group.bench_function("glsl", |b| {
+        let vert = gather_inputs("tests/in/glsl", "vert");
+        b.iter(move || parse_glsl(naga::ShaderStage::Vertex, &vert));
+        let frag = gather_inputs("tests/in/glsl", "frag");
+        b.iter(move || parse_glsl(naga::ShaderStage::Vertex, &frag));
+        //TODO: hangs for some reason!
+        //let comp = gather_inputs("tests/in/glsl", "comp");
+        //b.iter(move || parse_glsl(naga::ShaderStage::Compute, &comp));
+    });
+}
+
+#[cfg(feature = "wgsl-in")]
+fn gather_modules() -> Vec<naga::Module> {
+    let inputs = gather_inputs("tests/in", "wgsl");
+    let mut parser = naga::front::wgsl::Parser::new();
+    inputs
+        .iter()
+        .map(|input| {
+            let string = std::str::from_utf8(input).unwrap();
+            parser.parse(string).unwrap()
+        })
+        .collect()
+}
+#[cfg(not(feature = "wgsl-in"))]
+fn gather_modules() -> Vec<naga::Module> {
+    Vec::new()
+}
```

Comment:
It's a little confusing that when necessary features aren't available, we do still run benchmarks and get meaningless numbers. The loop over the modules is inside the `b.iter` call, not outside it. It would be clearer to somehow just not run benchmarks that require features we don't have.

Reply:
that's a bit of an edge case. We don't really care about the numbers when features aren't enabled. What I did was a minimal thing to make it compile for these cases